### ISSUE TITLE
[Suggestion] Make cache expire time configurable.

### ISSUE THREAD

**Issue by KurtzPT** [ZeroQI/Hama.bundle#311]
I noticed this being a problem on airing anime. The default expire time on AniDB files is too long, I usually have the episodes titles of the airing animes not updated because the cache is old (while the information was already updated on AniDB page).

Example: One Punch S2.

AniDB Cache:
```
=== AniDB.GetMetadata() =====================================================================================================================================
Source: anidb, AniDBid: 12430, Full AniDBids list: ['12430'], Active AniDBids list: ['12430']
[+] 12430: []
language_posters: ['en', 'ja']
--- 12430 ---------------------------------------------------------------------------------------------------------------------------------------------------
AniDBid: 12430, IsPrimary: True, url: http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=12430
--- 12430.series --------------------------------------------------------------------------------------------------------------------------------------------
common.LoadFile() - file cached - CacheTime: 'Mon May  6 14:05:38 2019', Limit: 'Wed May 15 10:27:28 2019', url: 'http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=12430', Filename: 'AniDB/xml/12430.xml' file_valid: 'True'
```

Episode 4 and 5 don't have the updated titles, while AniDB webpage already has them updated:

```
--- 12430.episodes ------------------------------------------------------------------------------------------------------------------------------------------
[ ] ep creators (creators tag): {'producers': ['Suzuki Tomohiro'], 'directors': ['Sakurai Chikara'], 'writers': ['Murata Yuusuke', 'One']}
[X] s1e  1 => s1e  1 air_date: 2019-04-10 language_rank: 0, title: "Return of the Hero"
 - [ ] summary: 
[X] s1e  2 => s1e  2 air_date: 2019-04-17 language_rank: 0, title: "Human Monster"
 - [ ] summary: 
[X] s1e  3 => s1e  3 air_date: 2019-04-24 language_rank: 0, title: "The Hunt Begins"
 - [ ] summary: 
[X] s1e  4 => s1e  4 air_date: 2019-05-01 language_rank: 0, title: "Episode 4"
 - [ ] summary: 
[X] s1e  5 => s1e  5 air_date: 2019-05-08 language_rank: 0, title: "Episode 5"
 - [ ] summary: 
```

If I'm correct the file was downloaded 6 May and it only will be updated (redownloaded) almost 10 days after that day... It's way too much time when we have airing anime.

My suggestion would be making this a configurable parameter on Hama settings, or at least reduce the cache expire time to half.

**Comment by EndOfLine369** [comment]
FYI, fixed logging. Test was correct. Logging was wrong.
AniDB is set for 6 days which should be fine for the weekly interval for new running anime.

**Comment by icemouton** [comment]
@KurtzPT curious as to which folder structure you went with for OPM/season2 .. I had a bit of trouble making it work correctly

**Comment by ZeroQI** [comment]
@EndOfLine369 Thanks for the fix
@KurtzPT Could you articulate why less than 6 days cache is required?

**Comment by KurtzPT** [comment]
@ZeroQI 

In the example above, it just happens the timing when the cache expired and the scan re-downloaded the AniDB information (Mon May  6 14:05:38 2019), there was no information of the English titles of the 4th episode onwards. 

You can check in AniDB when the English title was updated:

- Episode 4: https://anidb.net/perl-bin/animedb.pl?show=creq&do.showhistov=1&table=eptitletb&id=536825
- Episode 5: https://anidb.net/perl-bin/animedb.pl?show=creq&do.showhistov=1&table=eptitletb&id=541370

In my example you can see that when the episode airs, there's no information on AniDB because usually these titles are just added after being aired (and sometimes it takes 1-2 days). Also I'm almost sure that AniDB API has requests cached so when something changes, it takes a bit for that cached requests to clear and have the new information available via API, so keep that in mind.

So with 6 days of cache here + AniDB API own caching + the delay of having the information updated, all of this combined will make make having airing anime information updated quite hard.

TL:DR: It's been **10 days** after episode 4 of One Punch S2 aired, and I still don't have the title updated because of the huge cache time Hama has. (Title was added 4 days after airing).

It's not hard for myself to change the code and set it to 3 days instead of 6, I just don't think it's a well thought cache time for airing animes.

@icemouton 
I use for the folder name **One Punch Man (2019) [anidb-12430]**
But you have to rename the "S2" of the file names to something else, or else ASS/Hama will interpret it as special episode and it won't map to the correct episodes. I usually rename every "S2" to "2nd Season".

**Comment by ZeroQI** [comment]
if Plex doesn't refresh meta punctually (does a scan though) if no new files, we can be more aggressive
Timeline but we need to spare the source ressources
- J - Episode air
- J+2 - updated title in anidb
- J+4 - updated title refreshed in anidb cache ??
- J+11 title in Plex???

We need information on API caching.

We can do dynamic caching:
- If series in progress and last ep aired (not last listed) has a title we do not refresh
- If series in progress and last ep aired  (not last listed) has no title we refresh (cache 1 day)
- If series in finished and titles are present no refresh

That will not refresh old series un-necessarilly, which will make full meta refresh faster and withstand source downtime better. it will refresh in progress series if needed only but daily if title missing,

On the grander scale, it will lower API calls to source
In progress series may or may not be more likely to be cached and refreshed faster.

**Comment by KurtzPT** [comment]
I agree that with old series, the cache like it is is more than enough. 

My Plex setup, I have a folder with only airing animes, so I only scan that folder periodical to check new episodes (1 hour each scan). It avoids scanning old series and thus with this setup if the cache was lower or reworked somehow, it would diminish the chances of having information missing.

**Comment by LaaZa** [comment]
Had similar issue(or still have) where updated episode names weren't being pulled and just got the generic 'Episode 3' name. How would I go about forcing a cache update for this? I also have suggestion: because AniDB has generic 'Episode x' names, doesn't that mean that alternative names are never used as it thinks there is actually a name for it? So should hama regard 'Episode x' names as not set and try the alternatives.

**Comment by ZeroQI** [comment]
Depends on the language order and title source order.
Point taken on 'Episode X' to be consider empty for AniDB

**Comment by LaaZa** [comment]
What if cache management was more smart. Invalidating the cache every time an episode is set to air.

**Comment by ZeroQI** [comment]
Working on a rewriting of LoadFile() to do
- refresh cache if episode title not updated for series in progress once a day
- refresh cachenormal for in progress series
- refresh cache much slower for completed series
- save only if file correct to avoid issues
- delete corrupted file on disk

**Comment by ZeroQI** [comment]
Just wrote without testing. Need to go through the code but kept variables to a minimum and code simple...
https://gist.github.com/ZeroQI/557b22861064c97426abf51c426766f9
@EndOfLine369 here is my work in progress to avoid corrupted files and have smarter cache so series no longer airing update significantly less, will make the agent faster and lower the load on metadata sources

**Comment by ZeroQI** [comment]
Please test latest version and report, I did dynamic caching, which is different if the series is in progress or not.

**Comment by ZeroQI** [comment]
Any feedack at all?

**Comment by LaaZa** [comment]
Haven't had the time to test. I'll take a closer look next week. May be a bit slow to find issues though if you have to wait for new episodes to see if it works.
